db/db_impl_write.cc

```diff
@@ -722,6 +727,10 @@ Status DBImpl::PreprocessWrite(const WriteOptions& write_options,
     status = HandleWriteBufferFull(write_context);
   }
 
+  if (UNLIKELY(status.ok() && !trim_history_scheduler_.Empty())) {
+    status = TrimMemtableHistory();
```

Comment:
On second thought, can we trim history only after flush?

Reply:
you mean only call `TrimMemtableHistory()` when both `flush_scheduler_` and `trim_history_scheduler_` are not empty?

Reply:
I mean, can we do without the `trim_history_scheduler_`? In the past, we only trim history when we call `InstallMemtableFlushResults` after each flush. With this new option `max_memtable_size_to_maintain`, can we still (or, is it sufficient) perform history trimming in the same place?

Reply:
I think you are referring to memtable flush instead of trim history. `MemTableListVersion::memlist_` holds all immutable memtables yet to be flushed, `MemTableListVersion::memlist_history_` holds all flushed memtables waiting to be trimmed. Flush happens in `TryInstallMemtableFlushResults` and trim history happens in `MemTableListVersion::Add` and `MemTableListVersion::Remove`. 
About `trim_history_scheduler_`, it's basically a holder for cfds that requires trim history. While it can be merged with `flush_scheduler_` to maintain a simpler interface, we also need to pay a toll with more complicated logic involving `FlushScheduler`